repository/src/main/resources/db/changelog/db.changelog-rdm-6246.3.xml

```diff
@@ -0,0 +1,54 @@
+<!--
+Migrate webhooks from event into a new event_webhook table.
+There were previously three columns on event containing that event's webhooks.
+This takes a relational approach to enable better fetching strategies.
+-->
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog"
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd">
+    <changeSet id="rdm-6246.create_event_stage_enum" author="alex.mcausland@hmcts.net">
+        <!--This Enum models the different stages at which an event can fire a webhook-->
+        <sql>
+            create type event_stage as enum('START', 'PRE_SUBMIT', 'POST_SUBMIT');
+        </sql>
+        <rollback>drop type event_stage;</rollback>
+    </changeSet>
+    <changeSet id="rdm-6246.create_event_webhook_table" author="alex.mcausland@hmcts.net">
+        <!--This is a joining table between event and webhook-->
+        <sql>
+            create table event_webhook(
+                id serial unique not null,
+                webhook_id integer references webhook(id) not null,
+                event_id integer references event(id) not null,
+                event_stage event_stage not null,
+                /* An event can have one webhook per stage */
+                primary key(event_id, event_stage)
+            );
+        </sql>
+        <rollback>drop table event_webhook;</rollback>
+    </changeSet>
+    <changeSet id="rdm-6246.populate_event_webhook_table" author="alex.mcausland@hmcts.net">
+        <!--Populate the newly created table.-->
+        <!--For every event we need to insert a row for every start,-->
+        <!--presubmit and postsubmit webhook to link the event to its webhook-->
+        <sql>
+            insert into event_webhook(event_id, webhook_id, event_stage)
+            (
+                select id as event_id, webhook_start_id as webhook_id, 'START'::event_stage
+                from event
+                where webhook_start_id is not null
+                union select id, webhook_pre_submit_id, 'PRE_SUBMIT'
+                from event
+                where webhook_pre_submit_id is not null
+                union select id, webhook_post_submit_id, 'POST_SUBMIT'
+                from event
+                where webhook_post_submit_id is not null
+            );
+            <!--
+            Prompt Postgres to update its statistics for the table since we have inserted a lot of rows.
+            Otherwise Postgres may generate inefficient query plans until the next time its autovacuum daemon runs.
+            -->
+            analyze event_webhook;
```

Comment:
wonder whether it'll be safer to do all the db preparation and data migration outside of Def Store deployment in a preliminary stage. We could provide the migration scripts to our DBA Alliu and ask him to do the migration.That would give us more control? I'm a bit concerned on what might happen to the prod Db if there's some issue during the Def Store deployment while data migration is in progress. Maybe I'm being too cautious but let's have a chat about this.

Reply:
These migrations should be fail safe since they are nondestructive; the existing schema should be preserved so the current App code can continue to run during and after the migration (and can continue to run unaffected if any part of the migration fails).

For sure we cannot be too cautious, we should go into it in depth next week.

Reply:
what we could do is asking Alliu, our DBA, to do a full db backup just before the deployment. That should be enough I think. I added a comment on the ticket in this regard.